### DISCUSSION TITLE
Occaional media playback stutter

### DISCUSSION THREAD

**Discussion by MartinHemma** [SpiralLinux/SpiralLinux-project#493] [Q&A] [answered]
Sometime earlier this year I started to notice that my laptop running SpiralLinux stutters when watching youtube videos. First I thought Google was messing with me because I usually use Invidious. Hence, I didn't investigate and it only happens occasionally. Now, I have realized that this is not limited to online media playback. When this computer has a stutter episode it affects all media playback including music and video living on its SSD.

Computer vital statistics: Bog standard Lenovo X230 with newish battery and one year old SSD. Intel i5-3320M
OS: Debian 12, kernel 6.10.11+bpo-amd64

Observations: Stutter episodes are occasional, maybe once a day on average. Outside of these episodes everything runs smoothly. I have had this computer for several years and had no problems of this sort until earlier this year. Using btop I notice kworker is busy during stutter episodes. I do not know if the kworker activity is causing the stuttering or is just another thing triggered by whatever is causing the stuttering. I have noticed occasional higher than normal but less high kworker activity when there is no stuttering. CPU load is moderate (~10%) and there is room left in RAM during stutter episodes. Hence, I don't think HW limitations cause this.

Internet search: I see reports of Linux media playback stuttering being reported as far back as 10+ years ago but what I have found so far has not been very helpful.

Any ideas?

TIA

/Martin

**Comment by MartinHemma** [comment]
No-one?

/Martin

**Comment by marwenius** [comment]
Which browser are you using? Are there any extensions installed?

Have you tried a different browser?

**Reply by MartinHemma** [reply]
I am using Firefox.

Once in a stuttering episode all media playback stutters, not only browser based but also playing locally stored audio and video. Closing down Firefox does not help but maybe Firefox somehow triggers the stuttering episodes even-though it seems far fetched to me. There is sometimes days between stuttering episodes and I use Firefox all the time. I will test with some other browser for a while to see if it helps.

/Martin

**Reply by marwenius** [reply]
The X230 model is now at least 10 years old. Maybe just signs of age? In your case: Maybe replace the thermal paste or remove dust from the fan?

Or maybe it's because the graphics card (should be Intel HD Graphics 4000) is too old and is reaching its limits with more modern video codecs such as H.265/HEVC or VP9?

Maybe an extension like [h264ify](https://addons.mozilla.org/de/firefox/search/?q=h264ify) could be helpful?

**Reply by MartinHemma** [reply]
Ah, CPU thermal paste! Have been there done that! Maybe a year ago -- or was it this spring? The behavior caused by old thermal paste was different: Random and very sudden complete shut-downs.

During stutter episodes all media stutters -- even pure audio such as .ogg, .mp3 or .flac played through VLC or MPV.

When watching things using btop I see no signs of HW overload during stutter episodes. It does not seem to be load dependent as it happens randomly and as far as I can see not correlated to my usage.

I have started my trial with a different browser. Too early to draw any conclusions yet.

/Martin

**Reply by geckolinux** [reply]
It's a bit of a wild guess, but I wonder if the physical RAM is getting low and it's starting to swap to the zRAM and the zRAM compression algorithm is stealing CPU time from the media playback? Maybe try `sudo systemctl stop zramswap.service` and test?

**Reply by MartinHemma** [reply]
I will try this next time I experience stuttering.

CPU does not look overloaded though but maybe the media playback process gets low scheduling priority.

/Martin

**Reply by MartinHemma** [reply]
Sorry for  the wait but until today stutter episodes has happened when it has not been convenient to try out the effect of `sudo systemctl stop zramswap.service`. I just tried it and there was no change. Does this command stop zRAM compression instantly or does it have to run its course before stopping? I am thinking I might have been impatient and jumped to conclusion.

I have been using Brave rather than Firefox lately to see if that helps. Stutter episodes are _maybe_ less frequent but then there has also been at least one kernel update lately and there has always been a large dose of randomness to all this. Hence, no conclusion drawn regarding browser influence.

/Martin

**Reply by geckolinux** [reply]
Hi there, `sudo systemctl stop zramswap.service` immediately stops it.

```
$ free -m
               total        used        free      shared  buff/cache   available
Mem:           15623        4796        1888         721        9994       10826
Swap:          14842           9       14832

$ sudo systemctl stop zramswap.service 
   
$ free -m
               total        used        free      shared  buff/cache   available
Mem:           15623        4668        2015         729       10002       10954
Swap:              0           0           0
```

How about trying the Debian Stable kernel?

**Reply by MartinHemma** [reply]
I have tested with the 6.5.0-0.deb12.1-amd64 kernel and it is no better than the others.

/Martin

**Reply by marwenius** [reply]
> I have tested with the 6.5.0-0.deb12.1-amd64 kernel and it is no better than the others.
> 
> /Martin

Why the 6.5 kernel? Seems a bit arbitrary somehow.

If I were you, I think I would try it with a live system where not much is installed yet and see if the problems persist? So maybe a SpiralLinux live system or a live system based on Trixie?
- https://spirallinux.github.io/#download
- https://cdimage.debian.org/cdimage/weekly-live-builds/amd64/iso-hybrid/

Or if it is possible to install a completely fresh system on another SSD to see if there are any problems?

**Reply by MartinHemma** [reply]
Geckolinux suggested I test the Debian Stable kernel. Of the kernels listed in the GRUB menu 6.5.0-0.deb12.1-amd64 looked (to my untrained eyes) like a Debian Stable kernel.

Your live system idea is a good one but somewhat impractical as these stuttering episodes are very random events. The computer can run trouble free for days, sometimes even longer. Having to rely on a live system for that long...

/Martin

**Reply by geckolinux** [reply]
Hi Martin, the Debian Stable kernel is 6.1.x.

**Reply by marwenius** [reply]
It was a mistake on my part. I think 6.5.x was the backport kernel that shipped with the original SpiralLinux from 2023.

The stable kernel is 6.1.x. You can install it with:
```
sudo apt update
sudo apt install linux-image-amd64 -t stable
```

There you can see the different kernel versions:
https://packages.debian.org/search?keywords=linux-image-amd64&exact=1

To see which one you have installed on your system:
```
dpkg --list | grep -i "linux-image"
```

But the question is whether it is really the kernel. Perhaps it is simply a background process that runs at (ir)regular intervals, e.g. for updates or other computationally intensive tasks, and this leads to the performance drops that you are observing, which do not appear to be related to anything else.

**Reply by geckolinux** [reply]
> Perhaps it is simply a background process that runs at (ir)regular intervals, e.g. for updates or other computationally intensive tasks, and this leads to the performance drops that you are observing, which do not appear to be related to anything else.

This would be my guess too. Maybe the next time you notice a stutter you can run `systemctl list-timers` to see if there is a process whose execution time coincides with the stutter.

**Reply by MartinHemma** [reply]
`sudo apt update`
`sudo apt install linux-image-amd64 -t stable`

Points at version 6.12.9+bpo-amd64 which is already installed but can't be configured due to some dependency error. This is the situation on both my main computer and the X230. Version 6.1.0-21-amd64 is installed on my main system but not on my X230. I don't know why they are different. I have not intervened in kernel up-dates so far.

I will try out `systemctl list-timers` next. Neither cpu nor RAM seem to be overloaded during stutter episodes but then btop is sampling the situation with a 2 s frequency.

/Martin

**Reply by marwenius** [reply]
I think the -t option (short for --target-release) probably did not work because of the backport priority for the kernel (and other packages).

You could try to install the stable kernel version directly:
```
sudo apt update
sudo apt install linux-image-6.1.0-29-amd64
```
If the "stable-updates" repository is available, this could also work:
```
sudo apt update
sudo apt install linux-image-6.1.0-30-amd64
```

**Reply by MartinHemma** [reply]
- Running kernel version 6.1.0-30-amd64 does not help.
- One core is running at 11% capacity. All others run at single digit % load.
- 57% RAM used.
- `systemctl list-timers` tells me snapper-timeline.service ran 10 minutes ago.
- kworker/2:0-even uses 2.4% CPU and is the top load after media playback (music only, no video) stopped.

/Martin

**Reply by geckolinux** [reply]
> `systemctl list-timers` tells me snapper-timeline.service ran 10 minutes ago.

Did it also have playback stutter at that same time?

**Reply by MartinHemma** [reply]
Yes but I think this is a case of correlation but not causality. Snapper-timeline.service runs each hour on the hour and I have observed it twice more and then there was no hint of stutter. Something else is causing the stutter. Something less frequent and quite possibly random.

/Martin

**Reply by MartinHemma** [reply]
Maybe wrong-thinking but what about [tweaking kernel settings for low-latency](https://samiux.github.io/low_latency.html) or installing a low-latency kernel?

/Martin

**Reply by MartinHemma** [reply, marked as answer]
I decided to test the kernel settings for low-latency a month ago.

I have had no stutter episode since then. I can't say with absolute certainty the low-latency tweaking was the cure since I up-date the system all the time.

For now it seems this problem is history.

/Martin
